### PR TITLE
Fix Beam Search for CPU and CUDA... Also include it in our API

### PR DESCRIPTION
**Pull request by aciddelgado**
### Summary
Beam Search was hanging and not outputting correct results. Furthermore, it did not fit into our API design. This PR addresses the correctness and API problems with Beam search. We plan to improve both CPU and CUDA performance and memory efficiency in the near future.

### Issues Addressed
Here is a quick summary of the issues addressed by this PR... these apply to both CPU and CUDA implementations:
- No log-softmax normalization was performed before adding beam scores. This caused faulty outputs which did not match the ORT implementation.
- The `is_done` flag was not set or checked properly in the case of EOS token or `max_sequence_length`. This caused hanging, infinite looping, and memory buffer overflow. This sometimes gave the impression of bad performance, while in reality it was a correctness issue.
- `Finalize` was not called automatically. If a user didn't call it manually this could cause a floating point exception or other fault.
- There was no easy way to get output from Beam Search. `Finalize` was clunky and unintuitive as it didn't fit with our API.
- Our testing file was not up to date with our latest APIs.

### API Changes
Given the issues with `Finalize`, this PR introduces an update to the way Beam Search fits into our API. The user no longer has to manually call `Finalize` in order to access the Beam Search results. These are returned automatically by the `Generate()` function and can be accessed using batch beam indexing.

### CODE REVIEW

File: src/beam_search_scorer.h [hunk 1]
```diff
@@ -1,11 +1,13 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
-
+#include "sequences.h"
+#include <iostream>
```
Comment (yufenglee) [inline src/beam_search_scorer.h:4]: is the <iostream> required?

File: src/beam_search_scorer_cuda.cpp [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #include "search_cuda.h"
 #include "beam_search_scorer_cuda.cuh"
 #include "beam_search_scorer_cuda.h"
+#include <iostream>
```
Comment (yufenglee) [inline src/beam_search_scorer_cuda.cpp:6]: nit: remove the head file

File: src/beam_search_scorer_cuda.cpp [hunk 3]
```diff
@@ -72,15 +73,25 @@ void BeamSearchScorer_Cuda::Process(Sequences_Cuda& sequences,
...
+  cuda::LaunchBeamSearchScorer_Finalize(state_cpu_->batch_size_, *state_gpu_, sequences.GetSequences(), sequences.GetSequenceLength(), beam_hyps_, next_beam_scores_, stream_);
+}
+
+RoamingArray<int32_t> BeamSearchScorer_Cuda::GetBeamHypothesis(size_t batch_id, size_t beam_id) const {
+  cuda_host_unique_ptr<int32_t*> hypothesis_ptr = CudaMallocHostArray<int32_t*>(1);
+  cuda_host_unique_ptr<int> hypothesis_length = CudaMallocHostArray<int>(1);
+  cuda_host_unique_ptr<float> hypothesis_score = CudaMallocHostArray<float>(1);
+  cuda::LaunchBeamSearchScorer_GetHypothesisPtr(batch_id, beam_id, beam_hyps_, hypothesis_ptr.get(), hypothesis_length.get(), hypothesis_score.get(), stream_);
+  CudaCheck() == cudaStreamSynchronize(stream_);
+  // std::cout << "hypothesis length: " << *hypothesis_length.get() << std::endl;
```
Comment (yufenglee) [inline src/beam_search_scorer_cuda.cpp:92]: nit: remove this unnecessary std::cout. You can use the DumpTensor to dump the result, just like this:
https://github.com/microsoft/onnxruntime-genai/blob/00ceb80e1984c408459dfabe92a5b4eb97318578/src/models/model.cpp#L56-L60

File: src/search.cpp [hunk 4]
```diff
@@ -114,9 +147,22 @@ void BeamSearch_Cpu::SelectTop() {
   }
 
 #if 0
-  DumpMemory("Next Scores", next_scores);
-  DumpMemory("Next Tokens", next_tokens);
-  DumpMemory("Next Indices", next_indices);
+  // Print next_tokens, next_indices, next_scores
```
Comment (yufenglee) [inline src/search.cpp:150]: replace it with the DumpSpan: https://github.com/microsoft/onnxruntime-genai/blob/00ceb80e1984c408459dfabe92a5b4eb97318578/src/models/debugging.h#L9

File: src/search_cuda.cpp [hunk 5]
```diff
@@ -125,9 +133,30 @@ void BeamSearch_Cuda::SelectTop() {
   std::span<int32_t> next_indices{topk_next_indices_.get(), size};
 
 #if 0
-  DumpCudaMemory("Next Scores", next_scores);
-  DumpCudaMemory("Next Tokens", next_tokens);
-  DumpCudaMemory("Next Indices", next_indices);
+  // Copy next_tokens, next_indices, next_scores to CPU
```
Comment (yufenglee) [inline src/search_cuda.cpp:136]: replace it with DumpCudaSpan: https://github.com/microsoft/onnxruntime-genai/blob/00ceb80e1984c408459dfabe92a5b4eb97318578/src/models/debugging.h#L15

File: test/model_tests.cpp [hunk 6]
```diff
@@ -147,59 +147,59 @@ TEST(ModelTests, GreedySearchGptCuda) {
...
-    auto* expected_output_start = &expected_output[i * params->search.max_length];
-    EXPECT_TRUE(0 == std::memcmp(expected_output_start, sequence.data(), params->search.max_length * sizeof(int32_t)));
-  }
-}
-
-TEST(ModelTests, BeamSearchGptCuda) {
-  for (auto model_path : c_tiny_gpt2_model_paths)
-    Test_BeamSearch_Gpt_Cuda(model_path.first, model_path.second);
-}
+// void Test_BeamSearch_Gpt_Cuda(const char* model_path, const char* model_label) {
```
Comment (yufenglee) [inline test/model_tests.cpp:150]: Why comment them out?

**Comment by yufenglee** [comment]
> Beam Search was not working properly and had to be fixed. Furthermore, the way we accessed the output did not match our API.
> 
> This PR addresses the immediate problems with Beam search as soon as possible... further work can be done for better efficiency and testing.

Could you please add more details about the bug, e.g., what cause the BeamSearch not work properly.

File: test/model_tests.cpp [hunk 7]
```diff
@@ -83,26 +83,16 @@ TEST(ModelTests, BeamSearchGptFp32) {
...
-
-    search.SelectTop();
-  }
+  auto generator = Generators::CreateGenerator(*model, *params);
+  auto result = Generators::Generate(*model, *params);
 
-  std::vector<int32_t> output_sequence(static_cast<size_t>(search.params_->batch_size) * search.params_->search.max_length);
-  search.Finalize(1, Generators::cpu_span<int32_t>{output_sequence}, {});
+  // std::cout << "Expected output: " << tokenizer->Decode(expected_output) << "\r\n";
+  // std::cout << "Actual output: " << tokenizer->Decode(result[0]) << "\r\n";
```
Comment (baijumeswani) [inline test/model_tests.cpp:90]: Remove dead code.
```suggestion
```

File: test/model_tests.cpp [hunk 8]
```diff
@@ -174,24 +164,15 @@ void Test_BeamSearch_Gpt_Cuda(const char* model_path, const char* model_label) {
...
-
-  size_t sequence_length = params->batch_size * params->search.max_length;
-  auto output_sequence_cuda = Generators::CudaMallocArray<int32_t>(sequence_length);
-  auto output_sequence_cpu = std::make_unique<int32_t[]>(sequence_length);
-
-  generator->search_->Finalize(1, Generators::gpu_span<int32_t>(output_sequence_cuda.get(), sequence_length), {});
-  cudaMemcpyAsync(output_sequence_cpu.get(), output_sequence_cuda.get(), sequence_length * sizeof(int32_t), cudaMemcpyDeviceToHost, params->cuda_stream);
-  cudaStreamSynchronize(params->cuda_stream);
+  // std::cout << "Expected output: " << tokenizer->Decode(expected_output) << "\r\n";
+  // std::cout << "Actual output: " << tokenizer->Decode(result[0]) << "\r\n";
```
Comment (baijumeswani) [inline test/model_tests.cpp:170]:
```suggestion
```

File: src/search_cuda.cpp [hunk 9]
```diff
@@ -200,8 +208,23 @@ void BeamSearch_Cuda::AppendNextTokensToSequences() {
...
+  if (finalized_)
+    return;
+  beam_scorer_->Finalize(sequences_, num_return_sequences);
+  finalized_ = true;
+}
+
+RoamingArray<int32_t> BeamSearch_Cuda::GetSequence(size_t index) {
+  Finalize(params_->search.num_return_sequences);
+  size_t batch_id = index / params_->search.num_return_sequences;
+  size_t beam_id = index % params_->search.num_return_sequences;
```
Comment (baijumeswani) [inline src/search_cuda.cpp:221]: Use consts where possible.

File: src/search_cuda.cpp [hunk 10]
```diff
@@ -6,6 +6,7 @@
 #include "beam_search_topk.h"
 #include <queue>
 #include <random>
+#include <iostream>
```
Comment (baijumeswani) [inline src/search_cuda.cpp:9]: Remove iostream include if not needed.

File: src/search.cpp [hunk 11]
```diff
@@ -4,9 +4,36 @@
 #include "beam_search_scorer.h"
 #include <queue>
 #include <algorithm>
+#include <iostream>
```
Comment (baijumeswani) [inline src/search.cpp:7]: Remove iostream include if not needed.

File: src/beam_search_scorer_cuda.cpp [hunk 12]
```diff
@@ -72,15 +72,24 @@ void BeamSearchScorer_Cuda::Process(Sequences_Cuda& sequences,
 
 bool BeamSearchScorer_Cuda::IsDoneLater() const {
   cudaEventSynchronize(event_process_complete_);
+  // auto temp_state_gpu_ = CudaMallocHostArray<cuda::BeamScorerState>(1);
+  // cudaMemcpyAsync(temp_state_gpu_.get(), state_gpu_.get(), sizeof(cuda::BeamScorerState), ::cudaMemcpyDeviceToHost, stream_);
```
Comment (baijumeswani) [inline src/beam_search_scorer_cuda.cpp:76]: Can remove dead code if not needed.

File: src/search.cpp [hunk 13]
```diff
@@ -7,6 +7,32 @@
...
+  // Subtract max score and scale by temperature
+  std::transform(scores.begin(), scores.end(), scores.begin(), [max_score, temperature](float score) { return (score - max_score) / temperature; });
+
+  // Compute sum of exponentials
+  float const exp_sum = std::accumulate(scores.begin(), scores.end(), 0.0f, [](float a, float b) { return a + std::exp(b); });
+
+  // Subtract log of sum of exponentials from each score
+  std::transform(scores.begin(), scores.end(), scores.begin(), [exp_sum](float score) { return score - std::log(exp_sum); });
+}
+
```
Comment (yufenglee) [inline src/search.cpp:35]: nit: move this to the src/softmax.h:
https://github.com/microsoft/onnxruntime-genai/blob/616690213be076665f965842b01955c021f32334/src/softmax.h#L5

File: src/search.cpp [hunk 14]
```diff
@@ -62,10 +88,17 @@ int Search_Cpu::GetSequenceLength() const {
 }
 
 void BeamSearch_Cpu::SelectTop() {
+  // Normalize next token scores
+  for (int i = 0; i < params_->BatchBeamSize(); i++) {
+    std::span<float> const scores = next_token_scores_.subspan(static_cast<size_t>(i) * static_cast<size_t>(params_->vocab_size), params_->vocab_size);
+    LogSoftMax(scores, 1.0);  // Should this be log softmax?
```
Comment (yufenglee) [inline src/search.cpp:94]: yes, 1.0 is same as the LogSoftMax under src/softmax.h

File: src/search.cpp [hunk 15]
```diff
@@ -258,6 +278,17 @@ void GreedySearch_Cpu::AppendNextTokensToSequences() {
   }
 }
 
+bool BeamSearch_Cpu::IsDone() const {
+  if (beam_scorer_->IsDone()) {
+    return true;
+  } else if (sequences_.GetSequenceLength() == params_->search.max_length) {
+    if (g_log.enabled && g_log.hit_max_length)
+      Log("hit_max_length", "beam cuda hit");
```
Comment (yufenglee) [inline src/search.cpp:286]: what is "beam cuda hit"?

Comment (yufenglee) [reply src/search.cpp:286]: The first parameter needs to be label and 2nd parameter is the content: https://github.com/microsoft/onnxruntime-genai/blob/616690213be076665f965842b01955c021f32334/src/logging.cpp#L80

**Review by yufenglee — APPROVED** [review]

**Pull request merged**